### PR TITLE
Go micro performance. lexer atn config using murmur hash 32.

### PR DESCRIPTION
**Pull request by wxio**
The Go target needs some TLC at both a micro and macro level. A a micro level the lexer (and probably the parser) is slow, around 40x slower than a hand rolled one. Where the cpp target is only around 40% slower (see https://github.com/antlr/antlr4/issues/1540 & https://github.com/pboyer/antlr4/issues/92). At a macro level the LR optimisation (see https://github.com/antlr/antlr4/issues/192) have not been ported to the Go runtime.

This PR changed the LexerATNConfig hash code from a string that was hashed with fnv to hashcode ints that are murmur32 hashed. This gives about a 30% (benchmark number below) improvement (if implemented correctly), so there is still low hanging fruit in the 40x slow down. I intend to revisit this when I have some time or performance becomes an issue. My currently environment doesn't support profiling, which is probably an effective way of finding micro issues. Anyone interested in helping with this, I would be more than happy to collaborate.

@parrt and @sharwell any gotchas in implementing the LR changes?

**Comments**
**Benchmarking** - At this point in Antlr4's maturity it would be nice to have a benchmarking framework to complement  the unit testing framework. Go's benchmarking is an extension of testing and by the looks of it was influenced by [caliper](https://github.com/google/caliper), a Java benchmarking framework. Any suggestion of similar tools for the other targets and what such a framework might look like? Do any travis add-ins support reporting on benchmark data?

**Code coverage** - Whilst looking at the Go and Java code I came across another micro issue. In the Java runtime the `boolean add(T t)` in `Array2DHashSet` class will always return true (ie new element added), this behaviour also exists in the Go runtime. This doesn't effect behaviour, but without benchmarking it is difficult to tell the benefit from fixing it or replacing it with a hash set from a thirdparty library (eg [guava](https://github.com/google/guava) r19 is java 1.7 compatible). It wouldn't be a surprise if many micro issues exist across the runtime code bases. In this case the code in question dates back to 2012. So a deeper question is how to find them. The first answer that springs to mind is code coverage. Has anyone setup code coverage on Antlr4 using one of the travis add-ins? What other tools and techniques might be of interest?

Cheers
Gary

Go benchmarking numbers for this PR (see [modified dgraph benchmark](https://github.com/millergarym/dgraph/blob/bench-antlr4/antlr4go/lexer/benchmark_test.go))
@ashishne thought this might interest you.
before
```
bash-3.2$ go test -test.run=XXX -v  -benchtime=2s -bench . -cpuprofile cprof.prof
BenchmarkQuery/q1.1-12             10000            398395 ns/op
BenchmarkQuery/q1.2-12             10000            432751 ns/op
BenchmarkQuery/q2.1-12             10000            446644 ns/op
BenchmarkQuery/q2.2-12              5000            514398 ns/op
BenchmarkQuery/q3.1-12              5000            490114 ns/op
BenchmarkQuery/q3.2-12              5000            504340 ns/op
BenchmarkQuery/q4.1-12              5000            537611 ns/op
BenchmarkQuery/q4.2-12              5000            552284 ns/op
BenchmarkQuery/q5.1-12              5000            538648 ns/op
BenchmarkQuery/q5.2-12              5000            552259 ns/op
```

after
```
bash-3.2$ go test -test.run=XXX -v  -benchtime=2s -bench .
BenchmarkQuery/q1.1-12             10000            277898 ns/op
BenchmarkQuery/q1.2-12             10000            299240 ns/op
BenchmarkQuery/q2.1-12             10000            307816 ns/op
BenchmarkQuery/q2.2-12             10000            351049 ns/op
BenchmarkQuery/q3.1-12             10000            336241 ns/op
BenchmarkQuery/q3.2-12             10000            352299 ns/op
BenchmarkQuery/q4.1-12             10000            374569 ns/op
BenchmarkQuery/q4.2-12             10000            384309 ns/op
BenchmarkQuery/q5.1-12             10000            374470 ns/op
BenchmarkQuery/q5.2-12             10000            386929 ns/op
```

From https://github.com/antlr/antlr4/issues/1540
```
query$ gotb -test.run=XXX -benchtime=2s -v
BenchmarkQueryParse/spielberg:handwitten:-4               100000         46918 ns/op
BenchmarkQueryParse/spielberg:antlr:-4                      2000       1741364 ns/op
BenchmarkQueryParse/tomhanks:handwitten:-4                100000         25982 ns/op
BenchmarkQueryParse/tomhanks:antlr:-4                       2000       1654579 ns/op
BenchmarkQueryParse/nestedquery:handwritten:-4             30000         73053 ns/op
BenchmarkQueryParse/nestedquery:antlr:-4                    1000       3385005 ns/op
PASS
ok      github.com/dgraph-io/dgraph/query    21.922s
```

### CODE REVIEW

**Comment by parrt** [comment]
verrrry interesting. Can Go people from @antlr/antlr-targets take a look at this?

File: runtime/Go/antlr/prediction_context.go [hunk 1]
```diff
@@ -322,6 +340,26 @@ func (a *ArrayPredictionContext) equals(other PredictionContext) bool {
 	}
 }
 
+func (a *ArrayPredictionContext) HashCode() int {
+	if a.isEmpty() {
+		return 19 //TODO what to return here, some prime is probably ol
```
Comment (willfaught) [inline runtime/Go/antlr/prediction_context.go]: Incomplete comment

Comment (willfaught) [inline runtime/Go/antlr/prediction_context.go]: Where does 19 come from?

File: runtime/Go/antlr/prediction_context.go [hunk 2]
```diff
@@ -322,6 +340,26 @@ func (a *ArrayPredictionContext) equals(other PredictionContext) bool {
...
+	}
+	v := 0
+	for i := 0; i < len(a.returnStates); i++ {
+		if a.returnStates[i] == BasePredictionContextEmptyReturnState {
+			v += (i * 23) //TODO what #
```
Comment (willfaught) [inline runtime/Go/antlr/prediction_context.go]: 23?

Comment (willfaught) [inline runtime/Go/antlr/prediction_context.go]: Incomplete comment

File: runtime/Go/antlr/prediction_context.go [hunk 3]
```diff
@@ -322,6 +340,26 @@ func (a *ArrayPredictionContext) equals(other PredictionContext) bool {
...
+			continue
+		}
+		v += a.returnStates[i]
+		if a.parents[i] != nil {
+			v += (27 * a.parents[i].HashCode())
```
Comment (willfaught) [inline runtime/Go/antlr/prediction_context.go]: "

File: runtime/Go/antlr/prediction_context.go [hunk 4]
```diff
@@ -322,6 +340,26 @@ func (a *ArrayPredictionContext) equals(other PredictionContext) bool {
...
+		} else {
+			v += 31
```
Comment (willfaught) [inline runtime/Go/antlr/prediction_context.go]: "

File: runtime/Go/antlr/semantic_context.go [hunk 5]
```diff
@@ -109,6 +111,10 @@ func (p *Predicate) equals(other interface{}) bool {
 	}
 }
 
+func (p *Predicate) HashCode() int {
+	return p.ruleIndex*43 + p.predIndex*47
```
Comment (willfaught) [inline runtime/Go/antlr/semantic_context.go]: "

File: runtime/Go/antlr/semantic_context.go [hunk 6]
```diff
@@ -155,6 +161,10 @@ func (p *PrecedencePredicate) equals(other interface{}) bool {
 	}
 }
 
+func (p *PrecedencePredicate) HashCode() int {
+	return p.precedence * 51
```
Comment (willfaught) [inline runtime/Go/antlr/semantic_context.go]: "

File: runtime/Go/antlr/semantic_context.go [hunk 7]
```diff
@@ -295,6 +305,21 @@ func (a *AND) evalPrecedence(parser Recognizer, outerContext RuleContext) Semant
 	return result
 }
 
+func (a *AND) HashCode() int {
+	v := 37
```
Comment (willfaught) [inline runtime/Go/antlr/semantic_context.go]: "

File: runtime/Go/antlr/semantic_context.go [hunk 8]
```diff
@@ -295,6 +305,21 @@ func (a *AND) evalPrecedence(parser Recognizer, outerContext RuleContext) Semant
...
+	for _, o := range a.opnds {
+		v += o.HashCode()
+	}
+	return v
+}
+func (a *OR) HashCode() int {
+	v := 41
```
Comment (willfaught) [inline runtime/Go/antlr/semantic_context.go]: "

File: runtime/Go/antlr/token.go [hunk 9]
```diff
@@ -14,6 +14,10 @@ type TokenSourceCharStreamPair struct {
 	charStream  CharStream
 }
 
+func NewTokenSourceCharStreamPair(source TokenSource, stream CharStream) *TokenSourceCharStreamPair {
```
Comment (willfaught) [inline runtime/Go/antlr/token.go]: Why is this needed?

File: runtime/Go/antlr/utils.go [hunk 10]
```diff
@@ -49,23 +49,34 @@ func (s *IntStack) Push(e int) {
 }
 
 type Set struct {
-	data           map[string][]interface{}
-	hashFunction   func(interface{}) string
-	equalsFunction func(interface{}, interface{}) bool
+	setData map[int][]interface{}
+	// stringhashFunction func(interface{}) string
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Dead code

File: runtime/Go/antlr/utils.go [hunk 11]
```diff
@@ -49,23 +49,34 @@ func (s *IntStack) Push(e int) {
...
+	equalsFunction func(interface{}, interface{}) bool) *Set {
 
 	s := new(Set)
 
-	s.data = make(map[string][]interface{})
+	s.setData = make(map[int][]interface{})
 
-	if hashFunction == nil {
-		s.hashFunction = standardHashFunction
+	// could be nil
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Does this add anything that the following if statement doesn't?

File: runtime/Go/antlr/utils.go [hunk 12]
```diff
@@ -49,23 +49,34 @@ func (s *IntStack) Push(e int) {
...
+	if hashcodeFunction != nil {
+		s.hashcodeFunction = hashcodeFunction
 	} else {
-		s.hashFunction = hashFunction
+		s.hashcodeFunction = standardHashFunction
 	}
 
+	// if hashFunction == nil {
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Dead code

File: runtime/Go/antlr/utils.go [hunk 13]
```diff
@@ -87,11 +98,16 @@ func standardEqualsFunction(a interface{}, b interface{}) bool {
...
+	if h, ok := a.(HashCodeer); ok {
+		return h.HashCode()
+	}
 
-	if ok {
-		return h.Hash()
+	if h, ok := a.(Hasher); ok {
+		s := h.Hash()
+		ha := fnv.New32a()
+		ha.Write([]byte((s)))
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Extra parens

Can we replace s with the expression?

File: runtime/Go/antlr/utils.go [hunk 14]
```diff
@@ -110,48 +126,47 @@ func standardHashFunction(a interface{}) string {
 type Hasher interface {
 	Hash() string
 }
-
-func hashCode(s string) string {
-	h := fnv.New32a()
-	h.Write([]byte((s)))
-	return fmt.Sprint(h.Sum32())
+type HashCodeer interface {
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Should be HashCoder (like Writer).

File: runtime/Go/antlr/utils.go [hunk 15]
```diff
@@ -110,48 +126,47 @@ func standardHashFunction(a interface{}) string {
...
+		s.setData[key] = append(s.setData[key], value)
 		return value
 	}
 
-	s.data[key] = []interface{}{value}
+	v := make([]interface{}, 1, 10)
+	v[0] = value
+	s.setData[key] = v
+
+	// s.setData[key] = []interface{}{value}
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Dead

File: runtime/Go/antlr/utils.go [hunk 16]
```diff
@@ -110,48 +126,47 @@ func standardHashFunction(a interface{}) string {
...
 			}
 		}
 
-		s.data[key] = append(s.data[key], value)
+		s.setData[key] = append(s.setData[key], value)
 		return value
 	}
 
-	s.data[key] = []interface{}{value}
+	v := make([]interface{}, 1, 10)
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Why cap 10?

File: runtime/Go/antlr/utils.go [hunk 17]
```diff
@@ -164,10 +179,10 @@ func (s *Set) contains(value interface{}) bool {
 func (s *Set) values() []interface{} {
 	l := make([]interface{}, 0)
 
-	for key := range s.data {
-		if strings.Index(key, "hash_") == 0 {
-			l = append(l, s.data[key]...)
-		}
+	for key := range s.setData {
+		// if strings.Index(key, "hash_") == 0 {
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Dead

File: runtime/Go/antlr/utils.go [hunk 18]
```diff
@@ -385,3 +400,36 @@ func TitleCase(str string) string {
...
+	c1_32 = 0xCC9E2D51
+	c2_32 = 0x1B873593
+	n1_32 = 0xE6546B64
+)
+
+func initHash(seed int) int {
+	return seed
+}
+
+func update(h1 int, k1 int) int {
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: Can we combine these into `h1, k1 int`?

File: runtime/Go/antlr/utils.go [hunk 19]
```diff
@@ -385,3 +400,36 @@ func TitleCase(str string) string {
...
+	k1 *= c1_32
```
Comment (willfaught) [inline runtime/Go/antlr/utils.go]: So many unexplained things. Is there a murmur hash already implemented that we can use? Something like https://gowalker.org/github.com/spaolacci/murmur3? Otherwise, this will need a lot of comments to explain what's going on.

**Comment by willfaught** [comment]
Are both commits related? One is so small, and the other so large, it'd be nice if they could be squashed if so.

Can we simplify the "Go micro performance. Changed lexer atn config from Hash() string to …
…HashCode() int. Using murmur hash 32." commit summary to something like: "Change lexer ATN config hash to murmur 32"?

File: runtime/Go/antlr/token.go [hunk 20]
```diff
@@ -208,7 +208,7 @@ func (c *CommonToken) String() string {
 		ch = ""
 	}
 
-	return "[" + "channel:" + strconv.Itoa(c.channel) + " @" + strconv.Itoa(c.tokenIndex) + "," + strconv.Itoa(c.start) + ":" + strconv.Itoa(c.stop) + "='" +
```
Comment (millergarym) [inline runtime/Go/antlr/token.go]: Hey @willfaught. I changed this local and pushed it without realising that the format needs to be consistent across all targets. Therefore reverted it.

File: runtime/Go/antlr/token.go (hunk 9 above)
Comment (millergarym) [inline runtime/Go/antlr/token.go]: As far and I remember this is an old change in an already accepted PR. I'll check, a rebase might give ride of it. If memory servers this was need to enable a custom lexer (eg a block based lexer)

File: runtime-testsuite/test/org/antlr/v4/test/runtime/go/BaseGoTest.java [hunk 21]
```diff
@@ -165,6 +165,9 @@ private static void cacheGoRuntime(File tmpPackageDir) throws Exception {
 	}
 
 	private static void copyFile(File source, File dest) throws IOException {
+		if ( source.isDirectory() ) {
```
Comment (millergarym) [inline runtime-testsuite/test/org/antlr/v4/test/runtime/go/BaseGoTest.java]: In general can you point me at a style guide. The bigger style issue with this in the exist of paren around the test. To much swapping between Java and Go :-)

**Comment by millergarym** [comment]
If possible can some one with hashing / crypto experience to do a high level look at this. I'm not at all confident at the approach I've taken.

@parrt any suggestions as to who is expert in this area?

**Comment by davesisson** [comment]
I second the import an existing implementation approach.  To get you started I found this particular implementation (there's a murmur hash in there) but haven't vetted it for capability:

https://github.com/leemcloughlin/gofarmhash

File: runtime/Go/antlr/atn_state.go [hunk 22]
```diff
@@ -122,6 +123,10 @@ func (as *BaseATNState) SetNextTokenWithinRule(v *IntervalSet) {
 	as.NextTokenWithinRule = v
 }
 
+func (as *BaseATNState) HashCode() int {
+	return as.stateNumber
```
Comment (millergarym) [inline runtime/Go/antlr/atn_state.go]: See comment as the end of this thread.

File: runtime/Go/antlr/lexer_action.go [hunk 23]
```diff
@@ -51,6 +52,10 @@ func (b *BaseLexerAction) getIsPositionDependent() bool {
 	return b.isPositionDependent
 }
 
+func (b *BaseLexerAction) HashCode() int {
+	return b.actionType
```
Comment (millergarym) [inline runtime/Go/antlr/lexer_action.go]: ditto

File: runtime/Go/antlr/atn_config.go [hunk 24]
```diff
@@ -285,6 +285,24 @@ func (l *LexerATNConfig) equals(other interface{}) bool {
 	return l.BaseATNConfig.equals(othert.BaseATNConfig)
 }
 
+func (l *LexerATNConfig) HashCode() int {
+	var f int
+	if l.passedThroughNonGreedyDecision {
```
Comment (millergarym) [inline runtime/Go/antlr/atn_config.go]: Again, can you point me at a style guide. Vscode fmt's the code on every save. As for white space between lines inside blocks, imo they should be used really sparingly. Is there a style guide for the Java antlr code?

File: runtime/Go/antlr/lexer_action_executor.go [hunk 25]
```diff
@@ -30,8 +31,10 @@ func NewLexerActionExecutor(lexerActions []LexerAction) *LexerActionExecutor {
 	// of the performance-critical {@link LexerATNConfig//hashCode} operation.
 
 	var s string
+	l.cachedHash = initHash(57)
```
Comment (millergarym) [inline runtime/Go/antlr/lexer_action_executor.go]: random primes

File: runtime/Go/antlr/utils.go (hunk 16 above)
Comment (millergarym) [inline runtime/Go/antlr/utils.go]: This isn't a cap,  slice = append( slice, element) with expand the slice as needed.
It was an attempt at a performance optimisation. There are so few clashes in the lexer being benchmarked that it made no difference.

File: runtime/Go/antlr/utils.go (hunk 19 above)
Comment (millergarym) [inline runtime/Go/antlr/utils.go]: @willfaught, @davesisson 
There are no current dependancies in the runtime outside of the standard library. There should be a really good reason to introduce one. Murmur hash is small, not going chancing (at least not in this universe, but lets not get meta-physical) and was copied into the Java implementation.

Looking into using something like the guava collection in the Java code is an interesting discussion. Imo, importing murmur isn't.

**Comment by millergarym** [comment]
@willfaught thanks for the feedback, much appreciated. This PR was more about generating discussing than being merged and I've learnt a lot from your comments. Hopefully some hashing expert will provide some high level feedback. Time permitting I'll revisit this.

Any comments on including benchmarking as a standard part of the build?

Cheers
Gary

File: runtime/Go/antlr/lexer_atn_simulator.go [hunk 26]
```diff
@@ -311,7 +311,7 @@ func (l *LexerATNSimulator) accept(input CharStream, lexerActionExecutor *LexerA
 }
 
 func (l *LexerATNSimulator) getReachableTarget(trans Transition, t int) ATNState {
-	if trans.Matches(t, 0, LexerMaxCharValue) {
+	if trans.Matches(t, 0, 0xFFFE) {
```
Comment (parrt) [inline runtime/Go/antlr/lexer_atn_simulator.go]: @millergarym you need to pull from master as it looks like you are undoing our unicode 32 stuff.

File: runtime-testsuite/test/org/antlr/v4/test/runtime/go/BaseGoTest.java (hunk 21 above)
Comment (pboyer) [inline runtime-testsuite/test/org/antlr/v4/test/runtime/go/BaseGoTest.java]: Hey @millergarym, The best thing to do is to run `gofmt` on the file. BTW, in Go, you don't need parens around the conditional expression. `if expr { ... }` should work.

File: runtime/Go/antlr/atn_config.go (hunk 24 above)
Comment (pboyer) [inline runtime/Go/antlr/atn_config.go]: Again, just run `gotfmt` and we're good.

File: runtime/Go/antlr/utils.go (hunk 14 above)
Comment (pboyer) [inline runtime/Go/antlr/utils.go]: Yes, let's fix this typo.

File: runtime/Go/antlr/utils.go (hunk 19 above)
Comment (pboyer) [inline runtime/Go/antlr/utils.go]: @willfaught I'd prefer to copy paste an existing implementation, then to add a dependency for something as small as this.

**Comment by pboyer** [comment]
@millergarym Thanks, this is a good start. Do you plan to follow up on this? If not, I could probably start a new PR incorporating some of these ideas. You're right that there's a bunch of performance improvement to be done on the Go runtime.

**Comment by millergarym** [comment]
@pboyer, (cc: @parrt, @willfaught) lets close this PR and start a new one. I'm happy to help and think the best use of my time is looking into automated benchmarking. Travis doesn't seem to support this, however Jenkins does or can be made to via a combination of plugins. I'm happy to look into a Docker based Jenkins solution for non Travis CI stuff (benchmarking, cover coverage, todos and maybe even releases eg the XPath cyclic dependencies issue https://github.com/antlr/antlr4/issues/1620). Thoughts?
